README.md
CTF15
=====
In this era of digital world, there is one dark community rising with most intelligent individuals. Law maker call them criminals but normal people call them Hackers. If you are part of this community then this event is for you. Capture the flag is hacking challenge where you can prove your hacking skills. This challenge will test your understanding of web applications & operating systems.

CTF15 has two rounds.
---------------------
1) Online Qulification Round

2) Onsite Final Round

Team CTF:
========

Frontend Team Lead
------------------
Amar Sharma

Manvee

Rishita

Backend Team Lead
-----------------
Amar Sharma

Nishant Gandhi

Vartika Tewari

Posters Lead
------------
Kuntal Das

General Contributors
--------------------
Dipawesh Pawar

Gagan Kumar

Abhishek Jaiswal

Apoorva Shrivastava

Shailesh Kumar Kashyap

Ankit Choudhary

Rishabh Goel

Shifona Garg

Ankit Tripathi

Bhawana Pushkar 


CHANGELOG
Anwesha-CTF Change log

Release 2.0 - unreleased

#12 Front page of CTF ( Manvee and Rishita by amar-sharma )

#7 Rearrange CTF Project ( amar-sharma )

#1 New Database Design and implementation for mysql ( amar-sharma )

Release 1.0 - released

src/java/CTFCode/Login.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package CTFCode;

import java.io.IOException;
import java.io.PrintWriter;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.servlet.RequestDispatcher;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;

/**
 *
 * @author nishant
 */
@WebServlet(name = "Login", urlPatterns = {"/Login"})
public class Login extends HttpServlet {

    /**
     * Processes requests for both HTTP <code>GET</code> and <code>POST</code>
     * methods.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    protected void processRequest(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException, ClassNotFoundException, SQLException, NoSuchAlgorithmException {
        response.setContentType("text/html;charset=UTF-8");
        PrintWriter out = response.getWriter();

        String uname = null;
        String pass = null;

        if (!(request.getParameter("user") == null || request.getParameter("pass") == null)) {
            uname = request.getParameter("user");
            pass = request.getParameter("pass");
            System.out.println(uname+pass);

        } else {
            response.sendRedirect("index.html");
        }

        String plaintext = pass;
        MessageDigest m = MessageDigest.getInstance("MD5");
        m.reset();
        m.update(plaintext.getBytes());
        byte[] digest = m.digest();
        BigInteger bigInt = new BigInteger(1, digest);
        String hashtext = bigInt.toString(16);
// Now we need to zero pad it if you actually want the full 32 chars.
        while (hashtext.length() < 32) {
            hashtext = "0" + hashtext;
        }

        Class.forName("com.mysql.jdbc.Driver");
        Connection cn = DriverManager.getConnection("jdbc:mysql://localhost:3306/anwesha_ctf15", "ctf", "");
        PreparedStatement ps = cn.prepareCall("select tname,pass from login where tname=? and pass=?");
        ps.setString(1, uname);
        ps.setString(2, hashtext);
        ResultSet rs = ps.executeQuery();
        int problemstotal;
        out.print(rs.isBeforeFirst());

        if (rs.first()) {

            HttpSession ses = request.getSession(true);
            ses.setAttribute("uname", rs.getString(1));

            PreparedStatement total = cn.prepareCall("select total from problems where Id=?");
            total.setString(1, uname);
            ResultSet rstotal = total.executeQuery();
            rstotal.first();
            ses.setAttribute("score", rstotal.getInt("total"));
            total = cn.prepareCall("select * from problems where Id=?");
            total.setString(1, uname);
            rstotal = total.executeQuery();
            rstotal.first();

            problemstotal = rstotal.getInt("p1") + rstotal.getInt("p2") + rstotal.getInt("p3") + rstotal.getInt("p4") + rstotal.getInt("p5") + rstotal.getInt("p6") + rstotal.getInt("p7") + rstotal.getInt("p8") + rstotal.getInt("p9") + rstotal.getInt("p10") + rstotal.getInt("p11") + rstotal.getInt("p12") + rstotal.getInt("p13") + rstotal.getInt("p14") + rstotal.getInt("p15");

            ses.setAttribute("total", String.valueOf(problemstotal));

            response.sendRedirect("home.jsp");
        } else {
 //     RequestDispatcher rd = request.getRequestDispatcher("studentdetail.jsp?login=nologin");
            // rd.forward(request, response);
            // request.sendRedirect("");
            response.sendRedirect("index.jsp");
        }

        try {
            /* TODO output your page here. You may use following sample code. */
            out.println("<html>");
            out.println("<head>");
            out.println("<title>Servlet Login</title>");
            out.println("</head>");
            out.println("<body>");
            out.println("<h1>Servlet Login at " + request.getContextPath() + "</h1>");
            out.println("</body>");
            out.println("</html>");
        } finally {
            out.close();
        }
    }

    // <editor-fold defaultstate="collapsed" desc="HttpServlet methods. Click on the + sign on the left to edit the code.">
    /**
     * Handles the HTTP <code>GET</code> method.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        try {
            try {
                processRequest(request, response);
            } catch (NoSuchAlgorithmException ex) {
                Logger.getLogger(Login.class.getName()).log(Level.SEVERE, null, ex);
            }
        } catch (ClassNotFoundException ex) {
            Logger.getLogger(Login.class.getName()).log(Level.SEVERE, null, ex);
        } catch (SQLException ex) {
            Logger.getLogger(Login.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    /**
     * Handles the HTTP <code>POST</code> method.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        try {
            try {
                processRequest(request, response);
            } catch (NoSuchAlgorithmException ex) {
                Logger.getLogger(Login.class.getName()).log(Level.SEVERE, null, ex);
            }
        } catch (ClassNotFoundException ex) {
            Logger.getLogger(Login.class.getName()).log(Level.SEVERE, null, ex);
        } catch (SQLException ex) {
            Logger.getLogger(Login.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    /**
     * Returns a short description of the servlet.
     *
     * @return a String containing servlet description
     */
    @Override
    public String getServletInfo() {
        return "Short description";
    }// </editor-fold>
}


src/java/CTFCode/Registration.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package CTFCode;

import java.io.IOException;
import java.io.PrintWriter;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 *
 * @author nishant
 */
@WebServlet(name = "Registration", urlPatterns = {"/Registration"})
public class Registration extends HttpServlet {
    private PrintWriter out;
    /**
     * Processes requests for both HTTP <code>GET</code> and <code>POST</code>
     * methods.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    
    
    protected void processRequest(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException, NoSuchAlgorithmException, ClassNotFoundException, SQLException {
        response.setContentType("text/html;charset=UTF-8");
        out= response.getWriter();
        Class.forName("com.mysql.jdbc.Driver");
        Connection cn = DriverManager.getConnection("jdbc:mysql://localhost:3306/anwesha_ctf15", "ctf", "");
        String tname = request.getParameter("tname");
        String pass = request.getParameter("pass");
        String repass = request.getParameter("repass");
        String aid1 = request.getParameter("aid1");
        String aid2 = request.getParameter("aid2");
        String aid3 = request.getParameter("aid3");
        String phone = request.getParameter("phone");
        String clg = request.getParameter("college");
        String email = request.getParameter("email");
        out.write("input testing\n");
        
        if(!checkInputs(cn,response,tname,pass,repass,aid1,aid2,aid3,phone,clg,email))
        {
            out.close();
            return;
        }
        
        out.write("Inserting Data\n");
        
        String plaintext = pass;
        MessageDigest m = MessageDigest.getInstance("MD5");
        m.reset();
        m.update(plaintext.getBytes());
        byte[] digest = m.digest();
        BigInteger bigInt = new BigInteger(1, digest);
        String hashtext = bigInt.toString(16);
// Now we need to zero pad it if you actually want the full 32 chars.
        while (hashtext.length() < 32) {
            hashtext = "0" + hashtext;
        }

        PreparedStatement prst = cn.prepareCall("insert into user_info values(?,?,?,?,?,?,?)");
        prst.setString(1, tname);
        prst.setString(2, aid1);
        prst.setString(3, aid2);
        prst.setString(4, aid3);
        prst.setString(5, phone);
        prst.setString(6, clg);
        prst.setString(7, email);
        prst.execute();
        prst.close();
        PreparedStatement Auth = cn.prepareCall("insert into login values(?,?)");
        Auth.setString(1, tname);
        Auth.setString(2, hashtext);
        Auth.execute();
        Auth.close();
        PreparedStatement problem = cn.prepareCall("insert into user_track values(?,0,0,0,0,0,0,0,0,0,0)");
        problem.setString(1, tname);
        problem.execute();
        problem.close();
        cn.close();
        response.sendRedirect("index.jsp?info=done");

        try {
            /* TODO output your page here. You may use following sample code. */

        } finally {
            out.close();
        }
    }
    private boolean checkInputs(Connection cn,HttpServletResponse response,String tname,String pass,String repass,String aid1,String aid2,String aid3,String phone,String clg,String email) throws IOException, SQLException
    {
        if (tname.isEmpty() || pass.isEmpty() || aid1.isEmpty() || clg.isEmpty() || email.isEmpty() ) {
            out.write("incompleteForm\n");
            response.sendRedirect("index.jsp?prob=incomplete");
            return false;
        }
        
        if (!pass.equals(repass)) {
            out.write("passwordUnmatched\n");
            response.sendRedirect("index.jsp?prob=passUnMatch");
            return false;
        }
       
        if (tname.length() > 20 || pass.length() > 20 || aid1.length()>20||  email.length() > 50 || clg.length() > 50)
        {
            out.write("lengthShort\n");
            response.sendRedirect("index.jsp?prob=length");
            return false;
        }
        else if(((!aid2.isEmpty()) && aid2.length()>20) || ((!aid3.isEmpty()) && aid3.length()>20) )
        {
            out.write("lengthShort\n");
            response.sendRedirect("index.jsp?prob=length");
            return false;
        }
        if((!phone.isEmpty()))
        {
            if(phone.length()!=10)
            {
            out.write("phoneFormat\n");
            response.sendRedirect("index.jsp?prob=phone");
            return false;
            }
                  
        }    
        return checkDatabase(cn,response,tname,aid1,aid2,aid3,phone);
        
    }
    private boolean checkDatabase(Connection cn,HttpServletResponse response,String tname,String aid1,String aid2,String aid3,String phone) throws SQLException, IOException
    {
        PreparedStatement ps = cn.prepareCall("select tname from user_info where tname=?");
        ps.setString(1, tname);
        ResultSet rs = ps.executeQuery();
        if (rs.first()) {
            out.write("TeamNameExist");
            response.sendRedirect("index.jsp?prob=teamExist");
            
            return false;
        } 
        
        
        ps = cn.prepareCall("select member1,member2,member3 from user_info where member1=? OR member2=? OR member3=?");
        ps.setString(1, aid1);
        ps.setString(2, aid1);
        ps.setString(3, aid1);
        
        rs = ps.executeQuery();
        if (rs.first()) {
            out.write("Dupilicate Entry!");
            response.sendRedirect("index.jsp?prob=idExist&name="+aid1);
            return false;
        }
        
        if(!aid2.isEmpty())
        {
            
            ps = cn.prepareCall("select member1,member2,member3 from user_info where member1=? OR member2=? OR member3=?");
            ps.setString(1, aid2);      
            ps.setString(2, aid2);      
            ps.setString(3, aid2);      
            rs = ps.executeQuery();
            if (rs.first()) {
                out.write("Dupilicate Entry!");
                response.sendRedirect("index.jsp?prob=idExist&name=" + aid2);
                return false;
            }

        }
        
        if(!aid3.isEmpty())
        {
            
            ps = cn.prepareCall("select member1,member2,member3 from user_info where member1=? OR member2=? OR member3=?");
            ps.setString(1, aid3);      
            ps.setString(2, aid3);      
            ps.setString(3, aid3);      
            
            rs = ps.executeQuery();
            if (rs.first()) {
                out.write("Dupilicate Entry!");
                response.sendRedirect("index.jsp?prob=idExist&name=" + aid3);
                return false;
            }

        }
        
        if(!aid3.isEmpty())
        {
            
            ps = cn.prepareCall("select member1,member2,member3 from user_info where member1=? OR member2=? OR member3=?");
            ps.setString(1, aid3);      
            ps.setString(2, aid3);      
            ps.setString(3, aid3);      
            
            rs = ps.executeQuery();
            if (rs.first()) {
                out.write("Dupilicate Entry!");
                response.sendRedirect("index.jsp?prob=idExist&name=" + aid3);
                return false;
            }

        }
        
        return true;
        
    }
    // <editor-fold defaultstate="collapsed" desc="HttpServlet methods. Click on the + sign on the left to edit the code.">
    /**
     * Handles the HTTP <code>GET</code> method.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        try {
            try {
                processRequest(request, response);
            } catch (ClassNotFoundException ex) {
                Logger.getLogger(Registration.class.getName()).log(Level.SEVERE, null, ex);
            } catch (SQLException ex) {
                Logger.getLogger(Registration.class.getName()).log(Level.SEVERE, null, ex);
            }
        } catch (NoSuchAlgorithmException ex) {
            Logger.getLogger(Registration.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    /**
     * Handles the HTTP <code>POST</code> method.
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        try {
            try {
                processRequest(request, response);
            } catch (ClassNotFoundException ex) {
                Logger.getLogger(Registration.class.getName()).log(Level.SEVERE, null, ex);
            } catch (SQLException ex) {
                Logger.getLogger(Registration.class.getName()).log(Level.SEVERE, null, ex);
            }
        } catch (NoSuchAlgorithmException ex) {
            Logger.getLogger(Registration.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    /**
     * Returns a short description of the servlet.
     *
     * @return a String containing servlet description
     */
    @Override
    public String getServletInfo() {
        return "Short description";
    }// </editor-fold>
}
